packages/kbn-openapi-bundler/src/bundler/process_document/document_processors/include_x_labels.ts

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import { logger } from '../../../logger';
+import { hasProp } from '../../../utils/has_prop';
+import { X_LABELS } from '../../known_custom_props';
+import { DocumentNode } from '../types/node';
+import { DocumentNodeProcessor } from './types/document_node_processor';
+
+/**
+ * Creates a node processor to exclude nodes having only one of provided tags.
+ */
+export function createIncludeXLabelsProcessor(labelsToInclude: string[]): DocumentNodeProcessor {
+  if (labelsToInclude.length === 0) {
+    throw new Error('"labelsToInclude" must have as minimum one label.');
+  }
+
+  const labelsToIncludeSet = new Set(labelsToInclude);
+  // Labels are only restricted to operations objects.
+  // An operation object must have `responses` field.
+  const canHaveLabels = (node: Readonly<DocumentNode>): boolean => {
+    return 'responses' in node;
+  };
+  const logUnsupportedNodeWarning = (location: string, xLabelsValue: unknown): void => {
+    const value = Array.isArray(xLabelsValue) ? xLabelsValue.join(', ') : xLabelsValue;
+
+    logger.warning(
+      `"${X_LABELS}: ${value}" in ${location} is ignored since "${X_LABELS}" is supported only for Operation objects.`
+    );
+  };
+
+  return {
+    shouldRemove(node, { parentKey }) {
+      const isValidNode = canHaveLabels(node);
+
+      if (!isValidNode && X_LABELS in node) {
+        logUnsupportedNodeWarning(parentKey.toString(), node[X_LABELS]);
+      }
+
+      if (!isValidNode || !(X_LABELS in node) || !Array.isArray(node[X_LABELS])) {
+        return false;
+      }
+
+      return node[X_LABELS].every((tag) => !labelsToIncludeSet.has(tag));
```

Comment:
It doesn't seem entirely correct. Some refactoring of the if checks reveals that we should exclude an operation if it has an invalid value of `x-labels`:

```suggestion
      if (!(X_LABELS in node)) {
        // No x-labels - no problem. The node should be included.
        return false;
      }

      const isValidNode = canNodeHaveLabels(node);
      const isValidLabels = Array.isArray(node[X_LABELS]);

      if (!isValidNode) {
        // x-labels can't be applied to this node.
        // We log a warning, but the node should still be included.
        logUnsupportedNodeWarning(parentKey.toString(), node[X_LABELS]);
        return false;
      }

      if (!isValidLabels) {
        // x-labels value is not valid.
        // We log a warning and remove the node because it doesnt contain the needed labels.
        logInvalidLabelsValueWarning(parentKey.toString(), node[X_LABELS]);
        return true;
      }

      return node[X_LABELS].every((label) => !labelsToIncludeSet.has(label));
```

Let's cover all the 3 edge cases in the e2e tests: `!(X_LABELS in node)`, `!isValidNode`, `!isValidLabels`.

In general, I'd recommend to always simplify your if statements and not mix more than 2 checks in a single `if`. The code complexity grows quickly, it becomes hard to read the code, but easy to miss an edge case.

Reply:
Thanks for finding this minor bug that operation objects with invalid `x-labels` (non array) should be skipped. I've added test scenarios to cover `!(X_LABELS in node)`, `!isValidNode`, `!isValidLabels`.

> In general, I'd recommend to always simplify your if statements and not mix more than 2 checks in a single if. The code complexity grows quickly, it becomes hard to read the code, but easy to miss an edge case.

I totally agree with that. The problem is that TS doesn't always provide this flexibility. Sometimes it requires to combine or inline type checks just to get a proper variable value inside/outside `if` code block. TS's type guards have limited usage. For example 

```ts
const isValidLabels = Array.isArray(node[X_LABELS]);

if (!isValidLabels) {
  return true;
}
```

isn't properly checked by TS so `node[X_LABELS]` after  `if (!isValidLabels)` block has `unknown` type. It requires to specify it in the `if` condition explicitly like

```ts
if (!Array.isArray(node[X_LABELS])) {
return true;
}
```

Reply:
@maximpn Thanks for triple-checking and fixing another edge case. The end result looks much better.